### PR TITLE
Make UsesGithubForDevelopment data provider smarter

### LINKED ISSUE #122: Make UsesGithubForDevelopment data provider smarter
An issue in #121 

We use [getMirrorUrl()](https://github-api.kohsuke.org/apidocs/org/kohsuke/github/GHRepository.html#getMirrorUrl()) to check if the given repository is an original repository or a mirror repository. 

```
if (repository.getMirrorUrl() == null || repository.getMirrorUrl().isEmpty()){
    System.out.print("Repository is a original repository");
}
System.out.print("Repository is a mirror repository");
```
But for mirror repos like [apache/tomee](https://github.com/apache/tomee). The `mirror_url` is `null`. Hence, the above check will fail. This issue can be seen in many mirror repos of [apache](https://github.com/apache). 

For reference: The `mirror_url` flag is `true` for [coreutils/coreutils](https://github.com/coreutils/coreutils)

Hence, find an alternative to confirm the verification if a repository is a mirror or an original repository.

### PR DESCRIPTION
**Pull request by artem-smotrakov**
Here is a list of updates:

- Updated the `UsesGithubForDevelopment` data provider to consider
  multiple facts about a project.
- Updated `UsesGithubForDevelopmentTest`.

This closes #122

### CODE REVIEW

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java [hunk 1]
```diff
@@ -39,31 +56,100 @@ protected Value fetchValueFor(GitHubProject project) throws IOException {
...
+   * @param commit The commit to be checked.
+   * @return True if user names from the commit can be resolved to existing GitHub users,
+   *         false otherwise.
    */
-  private boolean isMirror(GitHubProject project) throws IOException {
-    GHRepository repository = gitHubDataFetcher().repositoryFor(project, github);
-    return StringUtils.isNotEmpty(repository.getMirrorUrl());
+  private static boolean canResolveUsersFor(GHCommit commit) {
+    try {
+      commit.getCommitter();
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java:147]: I think we need a null check here and in the next line. What are we checking here is not clear to me.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java [hunk 2]
```diff
@@ -39,31 +56,100 @@ protected Value fetchValueFor(GitHubProject project) throws IOException {
...
+  boolean usesGitHubForDevelopment(
+      GHRepository repository, List<GHCommit> commits, double threshold) {
+
+    List<Predicate<GHRepository>> checks = Arrays.asList(
+
+        // check if the repository doesn't have an explicit link to the original repository
+        r -> StringUtils.isEmpty(r.getMirrorUrl()),
+
+        // check if the description mentions "mirror"
+        r -> !r.getDescription().toLowerCase().contains("mirror"),
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java:98]: Projects like 
- https://github.com/HannahMitt/HomeMirror
- https://github.com/MichMich/MagicMirror
would send the wrong information. But I guess the threshold is making sure that we do not look for a hard `true` to everything?

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java [hunk 3]
```diff
@@ -39,31 +56,100 @@ protected Value fetchValueFor(GitHubProject project) throws IOException {
...
    * 
-   * @param project of type {@link GitHubProject}.
-   * @return true if the project is a mirror project. Otherwise, false.
-   * @throws IOException may throw an exception during REST call to GitHub API.
+   * @param repository The project's repository.
+   * @param commits The project's commit history.
+   * @return True if it looks like that the project uses GitHub, false otherwise.
+   *
+   */
+  boolean usesGitHubForDevelopment(
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java:89]: I think this method could be static. As this is stateless.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java [hunk 4]
```diff
@@ -39,31 +56,100 @@ protected Value fetchValueFor(GitHubProject project) throws IOException {
...
+      GHRepository repository, List<GHCommit> commits, double threshold) {
+
+    List<Predicate<GHRepository>> checks = Arrays.asList(
```
Comment (sourabhsparkala) [inline src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java:92]: This arraylist can be static as well.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java (hunk 1 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java:147]: If a project doesn't use GitHub as the main development repository, then user ids in the commit messages are not necessary point to GitHub user ids. The `getCommitter()` and `getAuthor()` methods throw an i/o exception if they can't find a GitHub user for a username from the commit message.

I'll add a null check although it I am not sure if the GitHub data fercher can return nulls in the commit list. Anyway, it's better to have a check.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java (hunk 2 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java:98]: Correct.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java (hunk 3 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java:89]: Right. That would also simplify the test.

File: src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java (hunk 4 above)
Comment (artem-smotrakov) [reply src/main/java/com/sap/sgs/phosphor/fosstars/data/github/UsesGithubForDevelopment.java:92]: Not really at the moment. The lambdas in the array use the `commits` parameter but the `Predicate` interface takes only one parameter and it is `GHRepository`. To overcome this, I need to introduce a new interface for a check. The new interface should take both repository and list of commits as a parameter. At first, I thought using the `Predicate` interface is better (and shorter). I'll introduce a new interface for checks.

**Comment by artem-smotrakov** [comment]
> Minor comments.
> Everything else looks good

Thanks for the review @sourabhsparkala ! I've addressed the comments.

**Pull request merged**